### ISSUE TITLE
Infinite score

### ISSUE THREAD

**Issue by jialiwang1211** [privefl/bigsnpr#139]
Hi Florian, 

I used ldpred2 to compute PRS on my data (n~75k, p~500k). The scores were computed on each chromosome and then added. The prediction is pretty good using infinitesimal model (snp_ldpred2_inf) -- high AUC and R2. However, when using the grid search on the combination of h2_seq=[0.7,1,1.4]*h2_est, p_seq =seq_log(1e-4, 1, length.out = 17), sparse=[TRUE, FALSE], the scores tend to go to infinity with moderate to small p_seq (<0.5). 

Does it suggest the Gibbs sampler is not converged or something else went wrong? How am I able to estimate PRS with small fraction p? Thank you!

**Comment by privefl** [comment]
Could you produce the same plot as in the tutorial and show it to me please?

![image](https://user-images.githubusercontent.com/11161036/92484924-93a61d80-f1ea-11ea-9119-ae0c213933a9.png)

**Comment by jialiwang1211** [comment]
It seems that the grid model works well on some chromosomes but produces extremely large predictive values or NA on others. Below are two examples on chromosome 2 and 6. 
<img width="775" alt="Screen Shot 2020-09-11 at 9 21 52 AM" src="https://user-images.githubusercontent.com/50063242/92931413-6be4ce80-f411-11ea-954f-559ce29eb44e.png">
<img width="775" alt="Screen Shot 2020-09-11 at 9 23 33 AM" src="https://user-images.githubusercontent.com/50063242/92931416-6d15fb80-f411-11ea-8057-e5542cb3565b.png">

**Comment by privefl** [comment]
Is this some autoimmune disease? 
Are you sure of the heritability estimate from chr6? It seems to be very small compared to the Z-Score that you get.

**Comment by jialiwang1211** [comment]
When I used the window size as 3cM instead of 500 SNPs to compute correlation, the scores do not go to infinity any more. 
It is not autoimmune disease. The heritability was computed by ldsc. Actually, I have a question on combining scores from chromosomes. When running the grid search model, should I estimate heritability on each chromosome separately, and define 3 candidate heritability parameters for that chromosome? In that case, the heritability parameters differ across chromosomes. 

In addition, I cannot reproduce results by using sap_ldpred2_auto model. It shows me the error message: "could not find function "ldpred2_gibbs_auto".

Thanks!

**Comment by privefl** [comment]
- Yes, 3cM is the recommended value.

- Yes, you would get different parameters per chromosome. Actually, we might revise the recommendation of running LDpred2 per chromosome soon.

- You should update to the latest bigsnpr version as this error has been solved already.

**Comment by jialiwang1211** [comment]
Thank you! I will update the bigsnpr package and test the auto model on my data.

**Comment by jialiwang1211** [comment]
I have another question about matching variants in the genotype data and summary statistics. In the plink bim file, the 5th column is A1 (coded as 1 in genotype matrix) and 6th column is A2 (coded as 0 in genotype matrix). In the summary statistics file, I have "Effect allele" and "Other allele" and the corresponding effect. The effect allele does not always correspond to the 5th column in the bim file. Can ldpred2 take care of that when merging the files, or I have to flip the sign of the effects in the summary statistics? 

For example, the first 6 columns are from the bim file, and the last 5 columns are from my summary statistics file. As you can see, for "rs573687", A1 is the effect allele, and for "rs523096", A2 is the effect allele. 

  CHR,  SNP,       POSG,      POS,   A1,    A2,    Effect_allele, Other_allele, Effect,  StdErr,        P  
     9,  rs573687,     0, 22011642, A,     G,     A,                  G,                    -0.223, 0.0136,      4.03e-60  
     9, rs523096,     0, 22019129, G,     A,    A,                   G,                    0.224,  0.0131,        1.39e-65

**Comment by privefl** [comment]
This is done by `snp_match()`.

**Comment by jialiwang1211** [comment]
Thanks. Just want to make clear that "a0" and "a1" in map file and sumstats file does not need to be the same. snp_match can take care of the flipped variants if their order does not the same?

**Comment by privefl** [comment]
It will do some possible flipping per variant, and also some possible reverse effect (if the alleles are switched) and will multiply the corresponding `beta` by -1.

**Comment by jialiwang1211** [comment]
Thanks for clarifying!

**Comment by jialiwang1211** [comment]
Hi Florian, 
I was a bit confused about the performance evaluation using training/testing data. In the tutorial, you train the model on the entire dataset, and choose the hyper-parameter on the training set, and then evaluate the performance (AUC etc.) on the testing set. 

Wouldn't it be that the training is applied sorely on the training set where you choose the best hyperparamters? 

Sorry if I mis-understood the program.

**Comment by privefl** [comment]
Not sure I understand your question.

Here, the training set is in fact a validation set (used only to find the best hyper-parameters).
Then, once that's chosen (only one model remains), then this model can be evaluated in a separate test set.

**Comment by jialiwang1211** [comment]
For example, in the code from your tutorial, you train the model on the entire data (assuming "corr", "df_beta" are computed from G). Then the best hypterparameters are chosen using GLM Z score on ind.val samples. Lastly, the AUC is reported on ind.test samples. 
 
```
ind.val <- sample(nrow(G), 400)
ind.test <- setdiff(rows_along(G), ind.val)

#train the model
beta_grid <- snp_ldpred2_grid(corr, df_beta, params, ncores = NCORES)
pred_grid <- big_prodMat(G, beta_grid, ind.col = ind.chr2)

#chose the best hyper parameters on training data
params$score <- big_univLogReg(as_FBM(pred_grid[**ind.val**, ]), y[**ind.val**])$score

#evaluate performance on testing data
pred_nosp <- big_prodVec(G, best_grid_nosp, ind.row = **ind.test**, ind.col = ind.chr2)
AUCBoot(pred_nosp, y[**ind.test**])
```

Is there a way to select samples when training/testing the model?

**Comment by privefl** [comment]
There is no real training happening in LDpred.
It is more like a "projection" based on LD and some external summary statistics.
Here, the fake sumstats are a bit odd, but let's pretend it's external sumstats.

Training (and validation and testing) happens when you use the phenotypes (`y`).

So, here the training would correspond to performing the GWAS to get the sumstats.
Validation is computing Z-scores using `big_univLogReg` and choosing the best one.
Testing is computing the AUC for the one chosen model.

**Comment by jialiwang1211** [comment]
Sorry for not stating my question properly. 
I understand that the sumstats are from external source. I guess my question is, when you computed "beta_grid" for example, you used G matrix for all samples (ind.val & ind.test), not just the G matrix corresponding to ind.val?

**Comment by privefl** [comment]
You mean `pred_grid`, right?
I guess I wanted to not have to compute all these scores multiple times.
But actually, I'm not using `pred_grid[ind.test, ]` anywhere, I'm just recomputing the final score for the test set.
So, I might just modify the tutorial so that it is not ambiguous (note that I'm only using `pred_grid[ind.val, ]`).

Does this answer your question?

**Comment by jialiwang1211** [comment]
Thanks for answering my question. 

I just ran the "auto" model following your tutorial. If I understood it correctly, the prediction
 is computed by averaging the "good" (measured by mad(sc)) prediction values across the chains with different initial values for "p". What is the rationale of merging the results from multiple chains ?

**Comment by privefl** [comment]
Please look at the corresponding section in the paper and [the code from the paper](https://github.com/privefl/paper-ldpred2/blob/master/code/run-ldpred2.R#L107-L108).

Averaging multiple runs is generally better, for many applications.

**Comment by privefl** [comment]
Thank you for using LDpred2. Please note that we now recommend running LDpred2 **genome-wide** instead of per chromosome. The paper (preprint) and tutorial have been updated.

**Comment by jialiwang1211** [comment]
Thank you for the update! I will read the latest paper and go through the tutorial.
